File: create_pattern/FactoryPattern.js
// 抽象产品类 (在JavaScript中不强制使用抽象类)
class Animal {
    speak() {
      throw "Method 'speak()' must be implemented.";
    }
  }
  
  // 具体产品类
  class Dog extends Animal {
    speak() {
      return "Woof!";
    }
  }
  
  class Cat extends Animal {
    speak() {
      return "Meow!";
    }
  }
  
  // 工厂类
  class AnimalFactory {
    createAnimal(animalType) {
      if (animalType === "dog") {
        return new Dog();
      } else if (animalType === "cat") {
        return new Cat();
      }
      return null;
    }
  }
  
  // 使用工厂模式创建对象
  const factory = new AnimalFactory();
  
  const dog = factory.createAnimal("dog");
  console.log(dog.speak());  // 输出: Woof!
  
  const cat = factory.createAnimal("cat");
  console.log(cat.speak());  // 输出: Meow!
  

File: create_pattern/FactoryPattern.py
from abc import ABC, abstractmethod

# 抽象产品类
class Animal(ABC):
    @abstractmethod
    def speak(self):
        pass

# 具体产品类
class Dog(Animal):
    def speak(self):
        return "Woof!"

class Cat(Animal):
    def speak(self):
        return "Meow!"

# 工厂类
class AnimalFactory:
    def create_animal(self, animal_type):
        if animal_type == "dog":
            return Dog()
        elif animal_type == "cat":
            return Cat()
        else:
            return None

# 使用工厂模式创建对象
factory = AnimalFactory()

dog = factory.create_animal("dog")
print(dog.speak())  # 输出: Woof!

cat = factory.create_animal("cat")
print(cat.speak())  # 输出: Meow!